go/vt/sqlparser/ast.go

```diff
@@ -2817,6 +2817,7 @@ type (
 	}
 
 	CountStar struct {
+		_ bool // This makes sure that reference equality checks works as expected
```

Comment:
I don't get this 🤷‍♂️

Reply:
You're correct that this might seem a bit strange at first glance. It's a quirk of Go's handling of empty structs. In Go, two instances of an empty struct are considered identical, which can be problematic when using these as keys in maps. They would be treated as the same key and potentially lead to incorrect map behavior.

Here's a brief example:

```golang
func TestWeirdGo(t *testing.T) {
	type CountStar struct{}

	cs1 := &CountStar{}
	cs2 := &CountStar{}
	require.NotSame(t, cs1, cs2, "what the what!?")
}
```

In the above code, cs1 and cs2, despite being distinct variables, would be treated as the same object.

The solution we employed was to add a dummy field `_ bool` to the otherwise empty struct `CountStar`. This ensures that each instance of `CountStar` is treated as a separate object, even in the context of map keys, such as the semantic state.